README
NAME
    Ado::Plugin::Site - Manage your Sites.

SYNOPSIS
      # in $ENV{MOJO_HOME}/etc/ado.config
      plugins => {
        # other plugins here...
        'admin', #required!
        'site',
        # other plugins here...
      }

       # GO to http://your-domain/ado
       # Access /ado-domains and ado-sites from the main admin menu

DESCRIPTION
    Ado::Plugin::Site is an Ado plugin that provides user interface to
    manage your domains and sites in the control panel provided by
    Ado::Plugin::Admin.

    Note: This software is not functional yet!

METHODS
    Ado::Plugin::Site inherits all methods from Ado::Plugin and implements
    the following new ones.

  register
      $plugin->register(Ado->new);

    Register plugin in Ado application.

CONTINUOUS INTEGRATION
    We would like to know that our software is always in good health. We
    count on friendly developers and organizations to install and test it
    continuously.

    CPAN Testers Reports for Ado::PLugin::Site
    <http://www.cpantesters.org/distro/A/Ado-Plugin-Site.html>

    Travis-CI <https://travis-ci.org/kberov/Ado-Plugin-Site>

SEE ALSO
    Ado::Plugin::Admin, Ado::Control::Ado::Domains, Ado::Control::Ado::Pages
    Ado::Plugin, Mojolicious::Guides::Growing, Ado::Manual, Mojolicious,
    <http://mojolicio.us>.

AUTHOR
    Krasimir Berov

COPYRIGHT AND LICENSE
    Copyright 2015 Krasimir Berov.

    This program is free software, you can redistribute it and/or modify it
    under the terms of the GNU Lesser General Public License v3 (LGPL-3.0).
    You may copy, distribute and modify the software provided that
    modifications are open source. However, software that includes the
    license may release under a different license.

    See http://opensource.org/licenses/lgpl-3.0.html for more information.



Changes
Revision history for Ado-Plugin-Site

0.06 2015-04-07
  - Added diagnostics to randomly failing tests.
  - Added Perl 5.16 for testing in Travis CI.

0.05 2015-04-03
  - Experimented with testing strategies and SQLite database.
    Now using a temporary folder for bootstrapped database.
    Now tests should pass.
  - Upgraded to Ado 0.87 to use enhanced DBI error reporting
    during database bootstrapping.
  - Added '.travis.yml'.
  - Added section CONTINUOUS INTEGRATION to the cocumentation.

0.04 2015-04-03
  - Using ':memory:' database during tests to avoid
    file permission propblems. 

0.03 2015-04-01
  - Fixed JavaScript error in
    public/plugins/site/pages.js create_page().

0.02 2015-03-31
  - We have nice implementation of a page-tree at
    /ado-pages/list and nice and lame jQuery application
    for this screen.
  - We have draft implementation of /ado-pages/create/:pid.
  - Fixed copy-paste errors where wrongly refering to table pages
    instead to domains.
  - Fixed typos in documentation.

0.01 2015-03-30
  - Released first ALPHA version to CPAN.
  - Added links to /ado-domains and /ado-pages to main admin menu.
  - Refactored automatically generated controllers to actually work.
  - Added basic mandatory tests i.e. perl-critic.t, manifest.t
    pod.t, pod-coverage.t, plugin/site-00.t
  - Added needed files to release to CPAN
  - Generated plugin using command
    "ado generate adoplugin -n Site --crud -t 'domains,pages'"


Build.PL
use 5.014;
use strict;
use warnings FATAL => 'all';
use Ado::BuildPlugin;

my $builder = Ado::BuildPlugin->new(
  module_name        => 'Ado::Plugin::Site',
  license            => 'lgpl_3_0',
  dist_version_from  => 'lib/Ado/Plugin/Site.pm',
  create_readme      => 1,
  dist_author        => q{Krasimir Berov <berov@cpan.org>},
  configure_requires => {'Module::Build' => 0.42, Ado => 0},
  build_requires => {'Test::More' => 0},
  requires       => {
    Ado                  => '0.87',
    'Ado::Plugin::Admin' => '0.13'
  },
  recursive_test_files => 1,
  meta_merge           => {
    resources => {
      repository => "https://github.com/kberov/Ado-Plugin-Site",
      bugtracker => "https://github.com/kberov/Ado-Plugin-Site/issues",
    },
    keywords =>
      [qw/Ado::Plugin::Site Mojolicious realtime web ERP REST CMS enterprise/],
    no_index => {namespace => ['Ado::Model'],}
  },
  add_to_cleanup => ['Ado-Plugin-Site-*', '*.bak'],
);

$builder->create_build_script();


lib/Ado/Model/Pages.pm
package Ado::Model::Pages;    #A table/row class
use 5.010001;
use strict;
use warnings;
use utf8;
use parent qw(Ado::Model);

sub is_base_class { return 0 }
my $TABLE_NAME = 'pages';

sub TABLE       { return $TABLE_NAME }
sub PRIMARY_KEY { return 'id' }
my $COLUMNS = [
  'id',        'pid',         'domain_id', 'alias',
  'page_type', 'sorting',     'template',  'cache',
  'expiry',    'permissions', 'user_id',   'group_id',
  'tstamp',    'start',       'stop',      'published',
  'hidden',    'deleted',     'changed_by'
];

sub COLUMNS { return $COLUMNS }
my $ALIASES = {};

sub ALIASES { return $ALIASES }
my $CHECKS = {
  'expiry' => {
    'allow'   => qr/(?^x:^-?\d{1,11}$)/,
    'default' => '86400'
  },
  'user_id'   => {'allow' => qr/(?^x:^-?\d{1,11}$)/},
  'id'        => {'allow' => qr/(?^x:^-?\d{1,}$)/,},
  'published' => {
    'allow'   => qr/(?^x:^-?\d{1,1}$)/,
    'default' => '0'
  },
  'stop' => {
    'default' => '0',
    'allow'   => qr/(?^x:^-?\d{1,11}$)/,
  },
  'pid' => {
    'default'  => '0',
    'defined'  => 1,
    'allow'    => qr/(?^x:^-?\d{1,11}$)/,
    'required' => 1
  },
  'page_type' => {
    'allow'    => qr/(regular|root|folder)/x,
    'required' => 1,
    'defined'  => 1
  },
  'cache' => {
    'allow'   => qr/(?^x:^-?\d{1,1}$)/,
    'default' => '0'
  },
  'alias' => {
    'defined'  => 1,
    'allow'    => qr/^\w{3,50}$/,
    'required' => 1
  },
  'permissions' => {
    'default'  => '-rwxr-xr-xr',
    'defined'  => 1,
    'required' => 1
  },
  'changed_by' => {
    'allow'    => qr/(?^x:^-?\d{1,11}$)/,
    'required' => 1,
    'defined'  => 1
  },
  'sorting' => {
    'allow'   => qr/(?^x:^-?\d{1,11}$)/,
    'default' => '1'
  },
  'tstamp' => {
    'allow'    => qr/(?^x:^-?\d{1,11}$)/,
    'defined'  => 1,
    'required' => 1,
    'default'  => '0'
  },
  'hidden' => {
    'default' => '1',
    'allow'   => qr/(?^x:^-?\d{1,1}$)/,
  },
  'start' => {
    'default' => '0',
    'allow'   => qr/(?^x:^-?\d{1,11}$)/,
  },
  'group_id' => {
    'default' => '1',
    'allow'   => qr/(?^x:^-?\d{1,11}$)/
  },
  'deleted' => {
    'default' => '0',
    'allow'   => qr/(?^x:^-?\d{1,4}$)/,
  },
  'domain_id' => {
    'allow'   => qr/(?^x:^-?\d{1,11}$)/,
    'default' => 1
  },
  'template' => {
    'allow' => sub {"DUMMY"}
  }
};

sub CHECKS { return $CHECKS }

#returns a list of objects with $self->pid == $self->id
sub children {
  my $self  = shift;
  my $class = ref($self);

  #TODO: add more realistic conditions like permissions, published etc..
  state $SQL = $class->SQL('SELECT') . ' WHERE pid=?';
  return [$class->query($SQL, $self->id)];
}
__PACKAGE__->QUOTE_IDENTIFIERS(0);

#__PACKAGE__->BUILD;#build accessors during load

1;

=pod

=encoding utf8

=head1 NAME

A class for TABLE pages in schema main

=head1 SYNOPSIS

  Ado::Model::Pages->create(domain_id=>1,alias=>'home',page_type=>'root');


=head1 DESCRIPTION


=head1 COLUMNS

Each column from table C<pages> has an accessor method in this class.

=head2 id

=head2 pid

=head2 domain_id

=head2 alias

=head2 page_type

=head2 sorting

=head2 template

=head2 cache

=head2 expiry

=head2 permissions

=head2 user_id

=head2 group_id

=head2 tstamp

=head2 start

=head2 stop

=head2 published

=head2 hidden

=head2 deleted

=head2 changed_by

=head1 ALIASES

None.

=head1 METHODS

=head2 children

Returns a possibly empty ARRAYREF of L<Ado::Model::Pages> objects C<WHERE>
C<$self-E<gt>pid> == C<$self-E<gt>id>. Note that the result is not cached
and every call to L</children> will execute a new SQL query.

=head1 GENERATOR

L<DBIx::Simple::Class::Schema>

=head1 SEE ALSO

L<Ado::Model>, L<DBIx::Simple::Class>, L<DBIx::Simple::Class::Schema>


=head1 AUTHOR

Красимир Беров (Krasimir Berov)

=head1 COPYRIGHT AND LICENSE

Copyright 2013-2015 Красимир Беров (Krasimir Berov).

This program is free software, you can redistribute it and/or
modify it under the terms of the 
GNU Lesser General Public License v3 (LGPL-3.0).
You may copy, distribute and modify the software provided that 
modifications are open source. However, software that includes 
the license may release under a different license.

See http://opensource.org/licenses/lgpl-3.0.html for more information.

=cut


public/plugins/site/pages.js
// public/plugins/site/pages.js
// browser side functionality for Ado::Control::Pages - part of Ado::Plugin::Site
/**
  Dropdown for folder and root pages.
  Displays popup with items for adding children pages under a page.
 */
(function($) {
'use strict';
$('.page.item .button').popup({
  inline   : true,
  hoverable: true,
  position : 'bottom right',
  delay: {
    show: 300,
    hide: 800
  }
});

// Initialise dropdowns in forms
$('select.dropdown').dropdown();

//  Let us define links behavior for pages items

/**
  Opens the modal form for creating a page
*/

function create_page () {
  var $mod = $('#create_page'); 
  $mod.modal('show');
  //set the action of the form
  var $form = $('#create_page form');

  $form.attr('action',$(this).data('href'));
  //TODO: implement form validation using Semantic UI when documented properly
  $form.on('submit', function (e) {
    var form = e.target;

    if(!form.alias.value.match(/^[\w-]{3,50}$/)){
      $('.field',form).addClass('error');
      $(form.alias).attr('placeholder', 'PLease add a valid page alias!');
      return false;
    }
    //now we can send the form
    $.post(
      $form.attr('action'),
      $(form).serialize(),
      function success (data) {
        $mod.modal('hide');
        $('.menu a[href="/ado-pages"]').click();//dirty reload of pages tree
      }
    );
    $mod.modal('hide');
    return false;
  });

}
//show this modal when clicking on a "create" 
// link item in the popup menu for a page
$('#tab_body .page span.create').each(function(i){
  $(this).click(create_page);
});
// TODO: Set validation for create_page form

/**
  Opens the modal form for updating a page
*/
function update_page () {
  alert('not implemented!');
  //$('#update_page').modal('show');
}
//show this modal when clicking on a "update" 
// link item in the popup menu for a page
$('.page span.update.item').each(function(){$(this).click(update_page)});

/**
  Opens the modal form for deleting a page
*/
function delete_page () {
    alert('not implemented!');
  //$('#delete_page').modal('show');
}
//show this modal when clicking on a "delete" 
// link item in the popup menu for a page
$('.page span.delete.item').each(function(){$(this).click(delete_page)});

})(jQuery); //execute;


lib/Ado/Control/Ado/Domains.pm
package Ado::Control::Ado::Domains;
use Mojo::Base 'Ado::Control::Ado';

#Parameters to generate a model class on the fly
my %t_class_params = (
  namespace => 'Ado::Model',
  table     => 'domains',
  type      => 'TABLE'
);

# List resourses from table domains.
sub list {
  my $c = shift;
  state $table_class = Ado::Model->table_to_class(%t_class_params);
  $c->require_formats('json', 'html') || return;
  my $args = Params::Check::check(
    { limit => {
        allow => sub { $_[0] =~ /^\d+$/ ? 1 : ($_[0] = 20); }
      },
      offset => {
        allow => sub { $_[0] =~ /^\d+$/ ? 1 : defined($_[0] = 0); }
      },
    },
    { limit  => $c->req->param('limit')  || 20,
      offset => $c->req->param('offset') || 0,
    }
  );

  $c->res->headers->content_range(
    "domains $$args{offset}-${\($$args{limit} + $$args{offset})}/*");
  $c->debug("rendering json and html only [$$args{limit}, $$args{offset}]");

  #Used in template domains/list.html.ep
  $c->stash('table_class', $table_class);

  #content negotiation
  my $list = $c->list_for_json([$$args{limit}, $$args{offset}],
    [$table_class->select_range($$args{limit}, $$args{offset})]);
  return $c->respond_to(
    json => $list,
    html => {list => $list}
  );
}

# Creates a resource in table domains. A naive example.
sub create {
  my $c = shift;
  state $table_class = Ado::Model->table_to_class(%t_class_params);
  my $v = $c->validation;
  return $c->render unless $v->has_data;

  $v->required('title')->size(3, 50);
  $v->required('body')->size(3, 1 * 1024 * 1024);    #1MB
  my $res;
  eval {
    $res = $table_class->create(
      title    => $v->param('title'),
      body     => $v->param('body'),
      user_id  => $c->user->id,
      group_id => $c->user->group_id,
      deleted  => 0,

      #permissions => '-rwxr-xr-x',
    );
  } || $c->stash(error => $@);                       #very rude!!!
  $c->debug('$error:' . $c->stash('error')) if $c->stash('error');

  my $data = $res->data;

  return $c->respond_to(
    json => {data => $data},
    html => {data => $data}
  );
}

# Reads a resource from table domains. A naive example.
sub read {    ##no critic 'Subroutines::ProhibitBuiltinHomonyms'
  my $c = shift;
  state $table_class = Ado::Model->table_to_class(%t_class_params);

  #This could be validated by a stricter route
  my ($id) = $c->stash('id') =~ /(\d+)/;

  my $data = $table_class->find($id)->data;
  $c->debug('$data:' . $c->dumper($data));
  return $c->respond_to(
    json => {article => $data},
    html => {article => $data}
  );
}

# Updates a resource in table domains.
sub update {
  my $c = shift;
  state $table_class = Ado::Model->table_to_class(%t_class_params);
  my $v    = $c->validation;
  my ($id) = $c->stash('id') =~ /(\d+)/;
  my $res  = $table_class->find($id);
  $c->reply->not_found() unless $res->data;
  $c->debug('$data:' . $c->dumper($res->data));

  if ($v->has_data && $res->data) {
    $v->optional('title')->size(3, 50);
    $v->optional('body')->size(3, 1 * 1024 * 1024);    #1MB
    $res->title($v->param('title'))->body($v->param('body'))->update()
      unless $v->has_error;
  }
  my $data = $res->data;
  return $c->respond_to(
    json => {article => $data},
    html => {article => $data}
  );
}

# "Deletes" a resource from table domains.
sub delete {    ##no critic 'Subroutines::ProhibitBuiltinHomonyms'
  return shift->render(message => '"delete" is not implemented...');
}


1;

__END__

=encoding utf8

=head1 NAME

Ado::Control::Ado::Domains - a controller for resource /ado-domains.


=head1 SYNOPSIS


  # Start Ado!
  #in your browser go to
  http://your-host/ado-domains/list
  #or
  http://your-host/ado-domains
  #and
  http://your-host/ado-domains/update/$id
  #and
  http://your-host/ado-domains/create


=head1 ATTRIBUTES

L<Ado::Control::Ado::Domains> inherits all the attributes from
L<Ado::Control::Ado>.

=head1 METHODS/ACTIONS

L<Ado::Control::Ado::Domains> inherits all methods from
L<Ado::Control::Ado> and implements the following new ones.
   


=head2 create

Handles route C</ado-domains/create>.
Responds to HTTP methods GET and POST.
Supported MIME types are 'application/json' and 'text/html'.

=head2 list

Handles route C</ado-domains> and C</ado-domains-list>.
Responds to HTTP method GET.
Supported MIME types are 'application/json' and 'text/html'.

=head2 read

Handles route C</ado-domains/read/:id>.
Responds to HTTP method GET.
Supported MIME types are 'application/json' and 'text/html'.

=head2 update

Handles route C</ado-domains/update/:id>.
Responds to HTTP methods GET and PUT.
Supported MIME types are 'application/json' and 'text/html'.

=head2 delete

Handles route C</ado-domains/delete/:id>.
Responds to HTTP method GET and DELETE.
Supported MIME types are 'application/json' and 'text/html'.

=head1 SEE ALSO

L<Ado::Control::Ado::Pages>, L<Ado::Control::Ado>, L<Ado::Control>

=head1 AUTHOR

Красимир Беров (Krasimir Berov)

=head1 COPYRIGHT AND LICENSE

Copyright 2015 Красимир Беров (Krasimir Berov).

This program is free software, you can redistribute it and/or
modify it under the terms of the
GNU Lesser General Public License v3 (LGPL-3.0).
You may copy, distribute and modify the software provided that
modifications are open source. However, software that includes
the license may release under a different license.

See http://opensource.org/licenses/lgpl-3.0.html for more information.

=cut


lib/Ado/Plugin/Site.pm
package Ado::Plugin::Site;
use Mojo::Base 'Ado::Plugin';
File::Spec::Functions->import(qw(catfile catdir));

our $VERSION = '0.06';

sub register {
  my ($self, $app, $config) = shift->initialise(@_);
  $self->_init_tables_and_data($config);
  $self->_amend_admin_menu();
  return $self;
}

# Add menu items to the Admin menu in the Administration UI
sub _amend_admin_menu {
  my $self          = shift;
  my $menu          = $self->app->admin_menu;
  my $content_items = $menu->items->first(sub { $_[0]->title eq 'Content' })->items;
  unshift @$content_items,
    Ado::UI::Menu->new(title => 'Domains', icon => 'world', url => '/ado-domains');
  return 1;
}

# Make sure the tables exist and are populated.
sub _init_tables_and_data {
  my ($self, $conf) = @_;
  my $app = $self->app;
  my $dbh = $app->dbix->dbh;

  #dummy check
  my $need_to_load_schema_and_data =
    eval { $app->dbix->query('SELECT * FROM pages WHERE id>1')->hash } ? 0 : 1;
  $app->log->debug($app->dumper($need_to_load_schema_and_data));
  if ($need_to_load_schema_and_data) {
    $app->do_sql_file(catfile($self->config_dir, $conf->{site_schema_sql_file}));
    $app->do_sql_file(catfile($self->config_dir, $conf->{site_data_sql_file}));
  }
  return;
}

1;

__END__

=encoding utf8

=head1 NAME

Ado::Plugin::Site - Manage your Sites.

=head1 SYNOPSIS

  # in $ENV{MOJO_HOME}/etc/ado.config
  plugins => {
    # other plugins here...
    'admin', #required!
    'site',
    # other plugins here...
  }

   # GO to http://your-domain/ado
   # Access /ado-domains and ado-sites from the main admin menu

=head1 DESCRIPTION

L<Ado::Plugin::Site> is an L<Ado> plugin that provides user interface to manage
your domains and sites in the control panel provided by L<Ado::Plugin::Admin>.

B<Note:> This software is not functional yet!


=head1 METHODS

L<Ado::Plugin::Site> inherits all methods from
L<Ado::Plugin> and implements the following new ones.

=head2 register

  $plugin->register(Ado->new);

Register plugin in L<Ado> application.

=head1 CONTINUOUS INTEGRATION

We would like to know that our software is always in good health.
We count on friendly developers and organizations to install and test it continuously.

L<CPAN Testers Reports for Ado::PLugin::Site|http://www.cpantesters.org/distro/A/Ado-Plugin-Site.html>

L<Travis-CI|https://travis-ci.org/kberov/Ado-Plugin-Site> 


=begin html

 <a href="https://travis-ci.org/kberov/Ado-Plugin-Site"><img 
  src="https://travis-ci.org/kberov/Ado-Plugin-Site.svg?branch=master"
    ></a><br /><br />

=end html


=head1 SEE ALSO

L<Ado::Plugin::Admin>,
L<Ado::Control::Ado::Domains>, L<Ado::Control::Ado::Pages>
L<Ado::Plugin>, L<Mojolicious::Guides::Growing>,
L<Ado::Manual>, L<Mojolicious>,  L<http://mojolicio.us>.


=head1 AUTHOR

Krasimir Berov

=head1 COPYRIGHT AND LICENSE

Copyright 2015 Krasimir Berov.

This program is free software, you can redistribute it and/or
modify it under the terms of the 
GNU Lesser General Public License v3 (LGPL-3.0).
You may copy, distribute and modify the software provided that 
modifications are open source. However, software that includes 
the license may release under a different license.

See http://opensource.org/licenses/lgpl-3.0.html for more information.

=cut
